File: RAnalytics/Core/Private/_RAnalyticsExternalCollector.m
#import <RAnalytics/RAnalyticsEvent.h>
#import "_RAnalyticsExternalCollector.h"
#import "_RAnalyticsHelpers.h"
#import <RAnalytics/RAnalytics-Swift.h>

static NSString *const _RAnalyticsLoginStateKey = @"com.rakuten.esd.sdk.properties.analytics.loginInformation.loginState";
static NSString *const _RAnalyticsTrackingIdentifierKey = @"com.rakuten.esd.sdk.properties.analytics.loginInformation.trackingIdentifier";
static NSString *const _RAnalyticsUserIdentifierKey = @"com.rakuten.esd.sdk.properties.analytics.loginInformation.userIdentifier";
static NSString *const _RAnalyticsLoginMethodKey = @"com.rakuten.esd.sdk.properties.analytics.loginInformation.loginMethod";

static NSString *const _RAnalyticsNotificationBaseName = @"com.rakuten.esd.sdk.events";

@interface _RAnalyticsExternalCollector ()
/*
 * Mutable properties
 */
@property (nonatomic, readwrite, getter=isLoggedIn) BOOL   loggedIn;
@property (nonatomic, nullable, readwrite, copy) NSString *trackingIdentifier;
@property (nonatomic, readwrite) RAnalyticsLoginMethod  loginMethod;
@property (nonatomic, nullable, readwrite, copy) NSString *logoutMethod;

/*
 * Private properties
 */
@property (nonatomic) NSDictionary                        *cardInfoEventMapping;
@property (nonatomic) NSDictionary                        *discoverEventMapping;
@end

@implementation _RAnalyticsExternalCollector

+ (instancetype)sharedInstance
{
    static _RAnalyticsExternalCollector *instance = nil;
    static dispatch_once_t _RAnalyticsExternalCollectorOnceToken;
    dispatch_once(&_RAnalyticsExternalCollectorOnceToken, ^{
        instance = [self.alloc initInstance];
    });
    return instance;
}

- (void)dealloc
{
    [NSNotificationCenter.defaultCenter removeObserver:self];
}

- (instancetype)init
{
    [self doesNotRecognizeSelector:_cmd];
    __builtin_unreachable();
}

- (instancetype)initInstance
{
    if (self = [super init])
    {
        [self addLoginObservers];
        [self addLoginFailureObservers];
        [self addLogoutObservers];
        [self addDiscoverObservers];
        [self addSSODialogObservers];
        [self addCredentialsObservers];
        [self addCustomEventObserver];
        
        [self update];
    }
    return self;
}

#pragma mark - Notification observers

- (void)addLoginObservers
{
    for (NSString *event in @[@"password", @"one_tap", @"other"])
    {
        NSString *eventName = [NSString stringWithFormat:@"%@.login.%@", _RAnalyticsNotificationBaseName, event];
        [self addNotificationName:eventName selector:@selector(receiveLoginNotification:)];
    }
}

- (void)addLoginFailureObservers
{
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(receiveLoginFailureNotification:)
                                                 name:[_RAnalyticsNotificationBaseName stringByAppendingString:@".login.failure"]
                                               object:nil];
}

- (void)addLogoutObservers
{
    for (NSString *event in @[@"local", @"global"])
    {
        NSString *eventName = [NSString stringWithFormat:@"%@.logout.%@", _RAnalyticsNotificationBaseName, event];
        [self addNotificationName:eventName selector:@selector(receiveLogoutNotification:)];
    }
}

- (void)addDiscoverObservers
{
    NSString *eventBase = [NSString stringWithFormat:@"%@.discover.", _RAnalyticsNotificationBaseName];
    
    _discoverEventMapping =  @{
                              @"visitPreview"           : NSNotification.discoverPreviewVisit,
                              @"tapPreview"             : NSNotification.discoverPreviewTap,
                              @"redirectPreview"        : NSNotification.discoverPreviewRedirect,
                              @"tapShowMore"            : NSNotification.discoverPreviewShowMore,
                              @"visitPage"              : NSNotification.discoverPageVisit,
                              @"tapPage"                : NSNotification.discoverPageTap,
                              @"redirectPage"           : NSNotification.discoverPageRedirect
                              };
    
    for (NSString *notification in _discoverEventMapping)
    {
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(receiveDiscoverNotification:)
                                                     name:[NSString stringWithFormat:@"%@%@", eventBase, notification]
                                                   object:nil];
    }
}

- (void)addSSODialogObservers
{
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(receiveSSODialogNotification:)
                                                 name:[_RAnalyticsNotificationBaseName stringByAppendingString:@".ssodialog"]
                                               object:nil];
}

- (void)addCredentialsObservers
{
    for (NSString *notification in
         @[@"ssocredentialfound",
           @"logincredentialfound"])
    {
        NSString *eventName = [NSString stringWithFormat:@"%@.%@", _RAnalyticsNotificationBaseName, notification];
        [self addNotificationName:eventName selector:@selector(receiveCredentialsNotification:)];
    }
}

- (void)addCustomEventObserver
{
    NSString *eventName = [_RAnalyticsNotificationBaseName stringByAppendingString:@".custom"];
    [self addNotificationName:eventName selector:@selector(receiveCustomEventNotification:)];
}

- (void)addNotificationName:(NSString *)name selector:(SEL)aSelector
{
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:aSelector
                                                 name:name
                                               object:nil];
}

#pragma mark - Handle notifications

- (void)receiveLoginNotification:(NSNotification *)notification
{
    [self update];
    
    if ([notification.object isKindOfClass:[NSString class]])
    {
        NSString *trackingIdentifier = [notification object];
        self.trackingIdentifier = trackingIdentifier;
    }
    
    self.loggedIn = YES;

    // For login we want to provide the logged-in state with each event, and each event tracker can know how the user logged in, so the loginMethod should be persisted.
    
    NSString *base = [NSString stringWithFormat:@"%@.login.", _RAnalyticsNotificationBaseName];
    if ([notification.name isEqualToString:[NSString stringWithFormat:@"%@password", base]])
    {
        self.loginMethod = RAnalyticsPasswordInputLoginMethod;
    }
    else if ([notification.name isEqualToString:[NSString stringWithFormat:@"%@one_tap", base]])
    {
        self.loginMethod = RAnalyticsOneTapLoginLoginMethod;
    }
    else
    {
        self.loginMethod = RAnalyticsOtherLoginMethod;
    }
    [self.class trackEvent:RAnalyticsLoginEventName];
}
- (void)receiveLoginFailureNotification:(NSNotification *)notification
{
    [self update];
    if ([notification.name isEqualToString:[_RAnalyticsNotificationBaseName stringByAppendingString:@".login.failure"]])
    {
        self.loggedIn = NO;
        self.trackingIdentifier = nil;
        NSMutableDictionary *parameters = [NSMutableDictionary new];
        
        if([notification.object isKindOfClass:[NSDictionary class]])
        {
            parameters[@"rae_error"] = notification.object[@"rae_error"];
            parameters[@"type"] = notification.object[@"type"];
            if (notification.object[@"rae_error_message"])
            {
                parameters[@"rae_error_message"] = notification.object[@"rae_error_message"];
            }
        }
        
        [self.class trackEvent:RAnalyticsLoginFailureEventName parameters:parameters.count ? parameters : nil];
    }
}

- (void)receiveLogoutNotification:(NSNotification *)notification
{
    [self update];
    self.loggedIn = NO;
    self.trackingIdentifier = nil;
    NSMutableDictionary *params = [NSMutableDictionary dictionary];
    if ([notification.name isEqualToString:[NSString stringWithFormat:@"%@.logout.local", _RAnalyticsNotificationBaseName]])
    {
        params[RAnalyticsLogoutMethodEventParameter] = RAnalyticsLocalLogoutMethod;
    }
    else
    {
        params[RAnalyticsLogoutMethodEventParameter] = RAnalyticsGlobalLogoutMethod;
    }
    [self.class trackEvent:RAnalyticsLogoutEventName parameters:params.copy];
}

- (void)receiveCardInfoNotification:(NSNotification *)notification
{
    NSString *key = [notification.name substringFromIndex:[NSString stringWithFormat:@"%@.cardinfo.", _RAnalyticsNotificationBaseName].length];
    [self.class trackEvent:_cardInfoEventMapping[key]];
}

- (void)receiveDiscoverNotification:(NSNotification *)notification
{
    NSString *eventSuffix = [notification.name substringFromIndex:[NSString stringWithFormat:@"%@.discover.", _RAnalyticsNotificationBaseName].length];
    
    NSArray *eventsRequiringOnlyIdentifier = @[@"tapPage", @"tapPreview"];
    NSArray *eventsRequiringIdentifierAndRedirectString = @[@"redirectPage", @"redirectPreview"];
    
    NSMutableDictionary *parameters = [NSMutableDictionary dictionaryWithCapacity:2];
    
    if ([eventsRequiringIdentifierAndRedirectString containsObject:eventSuffix] &&
             [notification.object isKindOfClass:NSDictionary.class] &&
             [notification.object[@"identifier"] isKindOfClass:NSString.class] &&
             [notification.object[@"url"] isKindOfClass:NSString.class])
    {
        parameters[@"prApp"] = notification.object[@"identifier"];
        parameters[@"prStoreUrl"] = notification.object[@"url"];
    }
    else if ([eventsRequiringOnlyIdentifier containsObject:eventSuffix] &&
             [notification.object isKindOfClass:NSString.class] &&
             ((NSString *)notification.object).length)
    {
        parameters[@"prApp"] = notification.object;
    }
    
    [self.class trackEvent:_discoverEventMapping[eventSuffix] parameters:parameters.count ? parameters : nil];
}

- (void)receiveSSODialogNotification:(NSNotification *)notification
{
    NSString *pageIdentifier = nil;
    if ([notification.object isKindOfClass:[NSString class]])
    {
        pageIdentifier = [notification object];
    }

    NSMutableDictionary *parameters = NSMutableDictionary.new;
    if (pageIdentifier.length)
    {
        parameters[@"page_id"] = pageIdentifier;
    }
    [self.class trackEvent:RAnalyticsPageVisitEventName parameters:parameters];
}

- (void)receiveCredentialsNotification:(NSNotification *)notification
{
    NSString *eventName = nil;
    NSMutableDictionary *parameters = NSMutableDictionary.new;
    
    if ([notification.name isEqualToString:[_RAnalyticsNotificationBaseName stringByAppendingString:@".ssocredentialfound"]])
    {
        eventName = RAnalyticsSSOCredentialFoundEventName;
    }
    else
    {
        eventName = RAnalyticsLoginCredentialFoundEventName;
    }
    
    if ([notification.object isKindOfClass:NSDictionary.class] &&
        [notification.object[@"source"] isKindOfClass:NSString.class])
    {
        parameters[@"source"] = notification.object[@"source"];
    }
    
    [self.class trackEvent:eventName parameters:parameters];
}

- (void)receiveCustomEventNotification:(NSNotification *)notification
{
    NSDictionary *object = notification.object;
    if (![object isKindOfClass:NSDictionary.class]) return;

    [self.class trackEvent:RAnalyticsCustomEventName parameters:notification.object];
}

#pragma mark - store & retrieve login/logout state & tracking identifier.

- (void)update
{
    NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults];
    _loggedIn           = [defaults boolForKey:_RAnalyticsLoginStateKey];
    _loginMethod        = [(NSNumber *)[defaults objectForKey:_RAnalyticsLoginMethodKey] unsignedIntegerValue];
    _trackingIdentifier = [defaults stringForKey:_RAnalyticsTrackingIdentifierKey];
    _userIdentifier = [defaults stringForKey:_RAnalyticsUserIdentifierKey];
}

- (void)setLoggedIn:(BOOL)loggedIn
{
    if (loggedIn != _loggedIn)
    {
        _loggedIn = loggedIn;
        NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults];
        [defaults setBool:loggedIn forKey:_RAnalyticsLoginStateKey];
        [defaults synchronize];
    }
}

- (void)setUserIdentifier:(NSString *)userIdentifier
{
    if (_RAnalyticsObjectsEqual(userIdentifier, _userIdentifier))
    {
        return;
    }
    
    _userIdentifier = userIdentifier.copy;
    NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults];
    if (userIdentifier.length)
    {
        [defaults setObject:userIdentifier forKey:_RAnalyticsUserIdentifierKey];
    }
    else
    {
        [defaults removeObjectForKey:_RAnalyticsUserIdentifierKey];
    }
    [defaults synchronize];
}

- (void)setTrackingIdentifier:(NSString *)trackingIdentifier
{
    if (!_RAnalyticsObjectsEqual(trackingIdentifier, _trackingIdentifier))
    {
        _trackingIdentifier = trackingIdentifier.copy;
        NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults];
        if (trackingIdentifier.length)
        {
            [defaults setObject:trackingIdentifier forKey:_RAnalyticsTrackingIdentifierKey];
        }
        else
        {
            [defaults removeObjectForKey:_RAnalyticsTrackingIdentifierKey];
        }
        [defaults synchronize];
    }
}

- (void)setLoginMethod:(RAnalyticsLoginMethod)loginMethod
{
    if (loginMethod != _loginMethod)
    {
        _loginMethod = loginMethod;
        NSUserDefaults* defaults = [NSUserDefaults standardUserDefaults];
        [defaults setObject:@(loginMethod) forKey:_RAnalyticsLoginMethodKey];
        [defaults synchronize];
    }
}

#pragma mark - Helpers

+ (void)trackEvent:(NSString *)eventName
{
    [self.class trackEvent:eventName parameters:nil];
}

+ (void)trackEvent:(NSString *)eventName parameters:(NSDictionary<NSString *, id> *)parameters
{
    [[RAnalyticsEvent.alloc initWithName:eventName parameters:parameters] track];
}
@end

File: RAnalytics/Core/Private/RAnalyticsExternalCollector.swift
import Foundation
import RLogger

@objc public protocol UserIdentifiable {
    var trackingIdentifier: String? { get }
    var userIdentifier: String? { get set }
}

/// This class tracks login, logout and push events.
/// It creates event corressponding to each event, sends it to RAnalyticsManager's instance to process.
@objc public final class RAnalyticsExternalCollector: NSObject, UserIdentifiable {
    private enum Constants {
        static let loginStateKey = "com.rakuten.esd.sdk.properties.analytics.loginInformation.loginState"
        static let trackingIdentifierKey = "com.rakuten.esd.sdk.properties.analytics.loginInformation.trackingIdentifier"
        static let userIdentifierKey = "com.rakuten.esd.sdk.properties.analytics.loginInformation.userIdentifier"
        static let loginMethodKey = "com.rakuten.esd.sdk.properties.analytics.loginInformation.loginMethod"
        static let notificationBaseName = "com.rakuten.esd.sdk.events"
    }

    /// The login state information is being stored in shared preferences.
    @objc public private(set) var isLoggedIn: Bool = false {
        willSet(newValue) {
            guard isLoggedIn != newValue else {
                return
            }
            userStorageHandler?.set(value: newValue, forKey: Constants.loginStateKey)
            userStorageHandler?.synchronize()
        }
    }

    /// The tracking identifier is being stored in shared preferences.
    @objc public private(set) var trackingIdentifier: String? {
        willSet(newValue) {
            guard trackingIdentifier != newValue else {
                return
            }
            if let trackingIdentifier = newValue,
               !trackingIdentifier.isEmpty {
                userStorageHandler?.set(value: trackingIdentifier, forKey: Constants.trackingIdentifierKey)

            } else {
                userStorageHandler?.removeObject(forKey: Constants.trackingIdentifierKey)
            }
            userStorageHandler?.synchronize()
        }
    }

    /// The user identifier is being stored in shared preferences.
    @objc public var userIdentifier: String? {
        willSet(newValue) {
            guard userIdentifier != newValue else {
                return
            }
            if let userIdentifier = newValue,
               !userIdentifier.isEmpty {
                userStorageHandler?.set(value: userIdentifier, forKey: Constants.userIdentifierKey)

            } else {
                userStorageHandler?.removeObject(forKey: Constants.userIdentifierKey)
            }
            userStorageHandler?.synchronize()
        }
    }

    /// The login method is being stored in shared preferences.
    @objc public private(set) var loginMethod: AnalyticsManager.State.LoginMethod = .other {
        willSet(newValue) {
            guard loginMethod != newValue else {
                return
            }
            // Note: loginMethod is set and got as NSNumber in RAnalytics version <= 7.x - _RAnalyticsExternalCollector.m
            userStorageHandler?.set(value: NSNumber(value: newValue.rawValue), forKey: Constants.loginMethodKey)
            userStorageHandler?.synchronize()
        }
    }

    /// Private

    private var logoutMethod: String?
    private var cardInfoEventMapping: NSDictionary?
    private let discoverEventMapping: [String: Notification.Name] = {
        ["visitPreview": .discoverPreviewVisit,
         "tapPreview": .discoverPreviewTap,
         "redirectPreview": .discoverPreviewRedirect,
         "tapShowMore": .discoverPreviewShowMore,
         "visitPage": .discoverPageVisit,
         "tapPage": .discoverPageTap,
         "redirectPage": .discoverPageRedirect]
    }()
    private let loginEvents = ["password", "one_tap", "other"]
    private let logoutEvents = ["local", "global"]
    private let credentialEvents = ["ssocredentialfound", "logincredentialfound"]
    private let eventsRequiringOnlyIdentifier = ["tapPage", "tapPreview"]
    private let eventsRequiringIdentifierAndRedirectString = ["redirectPage", "redirectPreview"]
    private let userStorageHandler: UserStorageHandleable?
    private let tracker: Trackable?

    @available(*, unavailable)
    override init() {
        self.userStorageHandler = nil
        self.tracker = nil
        super.init()
    }

    /// Creates an external collector
    ///
    /// - Parameters:
    ///   - dependenciesFactory: The dependencies factory.
    ///         It requires intances of these types: UserStorageHandleable, Trackable
    ///
    /// - Returns: An instance of RAnalyticsExternalCollector or nil.
    @objc public init?(dependenciesFactory: DependenciesFactory) {
        guard let userStorageHandler = dependenciesFactory.userStorageHandler,
              let tracker = dependenciesFactory.tracker else {
            RLogger.warning("RAnalyticsExternalCollector can't be created without userStorageHandler and tracker dependencies")
            return nil
        }
        self.userStorageHandler = userStorageHandler
        self.tracker = tracker
        super.init()
        addLoginObservers()
        addLoginFailureObservers()
        addLogoutObservers()
        addDiscoverObservers()
        addSSODialogObservers()
        addCredentialsObservers()
        addCustomEventObserver()
        update()
    }
}

// MARK: - Notification observers

private extension RAnalyticsExternalCollector {
    func addLoginObservers() {
        loginEvents.forEach {
            let evenName = "\(Constants.notificationBaseName).login.\($0)"
            addNotificationName(evenName, selector: #selector(receiveLoginNotification(_:)))
        }
    }

    func addLoginFailureObservers() {
        addNotificationName("\(Constants.notificationBaseName).login.failure", selector: #selector(receiveLoginFailureNotification(_:)))
    }

    func addLogoutObservers() {
        logoutEvents.forEach {
            let evenName = "\(Constants.notificationBaseName).logout.\($0)"
            addNotificationName(evenName, selector: #selector(receiveLogoutNotification(_:)))
        }
    }

    func addDiscoverObservers() {
        let eventBase = "\(Constants.notificationBaseName).discover."
        discoverEventMapping.forEach {
            addNotificationName("\(eventBase)\($0.key)", selector: #selector(receiveDiscoverNotification(_:)))
        }
    }

    func addSSODialogObservers() {
        let eventName = "\(Constants.notificationBaseName).ssodialog"
        addNotificationName(eventName, selector: #selector(receiveSSODialogNotification(_:)))
    }

    func addCredentialsObservers() {
        credentialEvents.forEach {
            let eventName = "\(Constants.notificationBaseName).\($0)"
            addNotificationName(eventName, selector: #selector(receiveCredentialsNotification(_:)))
        }
    }

    func addCustomEventObserver() {
        let eventName = "\(Constants.notificationBaseName).custom"
        addNotificationName(eventName, selector: #selector(receiveCustomEventNotification(_:)))
    }
}

// MARK: - Handle notifications

@objc private extension RAnalyticsExternalCollector {
    func receiveLoginNotification(_ notification: NSNotification) {
        update()

        if let trackingIdentifier = notification.object as? String {
            self.trackingIdentifier = trackingIdentifier
        }

        isLoggedIn = true

        // For login we want to provide the logged-in state with each event, and each event tracker can know how the user logged in, so the loginMethod should be persisted.

        let base = "\(Constants.notificationBaseName).login."
        switch notification.name.rawValue {
        case "\(base)password": loginMethod = .passwordInput
        case "\(base)one_tap": loginMethod = .oneTapLogin
        default: loginMethod = .other
        }

        tracker?.trackEvent(name: AnalyticsManager.Event.Name.login, parameters: nil)
    }

    func receiveLoginFailureNotification(_ notification: NSNotification) {
        update()

        guard notification.name.rawValue == "\(Constants.notificationBaseName).login.failure" else {
            return
        }

        isLoggedIn = false
        trackingIdentifier = nil

        var parameters = [String: Any]()

        if let params = notification.object as? [String: Any] {
            parameters["rae_error"] = params["rae_error"]
            parameters["type"] = params["type"]
            if let raeErrorMessage = params["rae_error_message"] {
                parameters["rae_error_message"] = raeErrorMessage
            }
        }

        tracker?.trackEvent(name: AnalyticsManager.Event.Name.loginFailure, parameters: parameters.isEmpty ? nil : parameters)
    }

    func receiveLogoutNotification(_ notification: NSNotification) {
        update()

        isLoggedIn = false
        trackingIdentifier = nil

        var parameters = [String: Any]()

        if notification.name.rawValue == "\(Constants.notificationBaseName).logout.local" {
            parameters[AnalyticsManager.Event.Parameter.logoutMethod] = AnalyticsManager.Event.LogoutMethod.local

        } else {
            parameters[AnalyticsManager.Event.Parameter.logoutMethod] = AnalyticsManager.Event.LogoutMethod.global
        }

        tracker?.trackEvent(name: AnalyticsManager.Event.Name.logout, parameters: parameters)
    }

    func receiveDiscoverNotification(_ notification: NSNotification) {
        let eventPrefix = "\(Constants.notificationBaseName).discover."
        let eventPrefixCount = eventPrefix.count
        guard eventPrefixCount < notification.name.rawValue.count else {
            return
        }

        let eventSuffix = notification.name.rawValue[eventPrefixCount..<notification.name.rawValue.count]

        var parameters = [String: Any]()

        if eventsRequiringIdentifierAndRedirectString.contains(eventSuffix),
           let dictionary = notification.object as? [String: Any],
           let identifier = dictionary["identifier"] as? String,
           let url = dictionary["url"] as? String {
            parameters["prApp"] = identifier
            parameters["prStoreUrl"] = url

        } else if eventsRequiringOnlyIdentifier.contains(eventSuffix),
                  let prApp = notification.object as? String,
                  !prApp.isEmpty {
            parameters["prApp"] = prApp
        }

        if let eventName = discoverEventMapping[eventSuffix]?.rawValue {
            tracker?.trackEvent(name: eventName, parameters: parameters.isEmpty ? nil : parameters)
        }
    }

    // Note: TO BE REMOVED?
    func receiveCardInfoNotification(_ notification: NSNotification) {
        let eventPrefix = "\(Constants.notificationBaseName).cardinfo."
        let eventPrefixCount = eventPrefix.count
        guard eventPrefixCount < notification.name.rawValue.count else {
            return
        }

        let key = notification.name.rawValue[eventPrefixCount..<notification.name.rawValue.count]
        if let eventName = cardInfoEventMapping?[key] as? String {
            tracker?.trackEvent(name: eventName, parameters: nil)
        }
    }

    func receiveSSODialogNotification(_ notification: NSNotification) {
        var pageIdentifier: String?
        if let aPageIdentifier = notification.object as? String {
            pageIdentifier = aPageIdentifier
        }
        var parameters = [String: Any]()
        if let pageIdentifier = pageIdentifier,
           !pageIdentifier.isEmpty {
            parameters["page_id"] = pageIdentifier
        }
        tracker?.trackEvent(name: AnalyticsManager.Event.Name.pageVisit, parameters: parameters)
    }

    func receiveCredentialsNotification(_ notification: NSNotification) {
        var eventName: String?
        if notification.name.rawValue == "\(Constants.notificationBaseName).ssocredentialfound" {
            eventName = AnalyticsManager.Event.Name.SSOCredentialFound

        } else {
            eventName = AnalyticsManager.Event.Name.loginCredentialFound
        }

        var parameters = [String: Any]()
        if let params = notification.object as? [String: Any],
           let source = params["source"] as? String {
            parameters["source"] = source
        }

        if let eventName = eventName {
            tracker?.trackEvent(name: eventName, parameters: parameters)
        }
    }

    func receiveCustomEventNotification(_ notification: NSNotification) {
        guard let parameters = notification.object as? [String: Any] else {
            return
        }
        tracker?.trackEvent(name: AnalyticsManager.Event.Name.custom, parameters: parameters)
    }
}

// MARK: - Store & retrieve login/logout state & tracking identifier.

private extension RAnalyticsExternalCollector {
    func update() {
        guard let userStorageHandler = userStorageHandler else {
            RLogger.warning("RAnalyticsExternalCollector can't be created without userStorageHandler dependency")
            return
        }
        isLoggedIn = userStorageHandler.bool(forKey: Constants.loginStateKey)
        trackingIdentifier = userStorageHandler.string(forKey: Constants.trackingIdentifierKey)
        userIdentifier = userStorageHandler.string(forKey: Constants.userIdentifierKey)

        // Note: loginMethod is set and got as NSNumber in RAnalytics version <= 7.x - _RAnalyticsExternalCollector.m
        let result = (userStorageHandler.object(forKey: Constants.loginMethodKey) as? NSNumber)?.uintValue ?? 0
        loginMethod = AnalyticsManager.State.LoginMethod(rawValue: result) ?? .other
    }
}

// MARK: - Helpers

private extension RAnalyticsExternalCollector {
    func addNotificationName(_ name: String, selector: Selector) {
        NotificationCenter.default.addObserver(self,
                                               selector: selector,
                                               name: NSNotification.Name(rawValue: name),
                                               object: nil)
    }
}
